### PR TITLE
TST: Test cleanup, parametrization for datetime64 arithmetic tests

### PR DESCRIPTION
**Pull request by jbrockmendel**
Avoids overlap with #23642.

### CODE REVIEW

File: pandas/tests/scalar/period/test_period.py [hunk 1]
```diff
@@ -509,35 +235,253 @@ def test_period_constructor_offsets(self):
...
+
+        msg = ('Frequency must be positive, because it'
+               ' represents span: -3M')
+        with pytest.raises(ValueError, match=msg):
+            Period('2011-01', freq='-3M')
+
+        msg = ('Frequency must be positive, because it' ' represents span: 0M')
+        with pytest.raises(ValueError, match=msg):
+            Period('2011-01', freq='0M')
+
```
Comment (jreback) [inline pandas/tests/scalar/period/test_period.py]: can parametrize next pass (and maybe split to 2 tests, working and error condtiions)

File: pandas/tests/scalar/period/test_period.py [hunk 2]
```diff
@@ -552,40 +496,79 @@ def test_hash(self):
...
+        assert p.tz == exp_zone.tzinfo
+        assert p.tz == exp.tz
 
-        with pytest.raises(period.IncompatibleFrequency):
-            left - Period('2007-01', freq='M')
+    @pytest.mark.parametrize('tzstr', ['dateutil/Europe/Brussels',
+                                       'dateutil/Asia/Tokyo',
+                                       'dateutil/US/Pacific'])
+    def test_to_timestamp_tz_arg_dateutil(self, tzstr):
+        from pandas._libs.tslibs.timezones import dateutil_gettz
```
Comment (jreback) [inline pandas/tests/scalar/period/test_period.py]: put the imports at the top

File: pandas/tests/scalar/period/test_period.py [hunk 3]
```diff
@@ -648,6 +631,103 @@ def _ex(p):
...
+        assert p.year == 1969
+        assert p.month == 11
+        assert isinstance(p, Period)
+
+    def test_freq_str(self):
+        i1 = Period('1982', freq='Min')
+        assert i1.freq == offsets.Minute()
+        assert i1.freqstr == 'T'
+
+    def test_period_deprecated_freq(self):
```
Comment (jreback) [inline pandas/tests/scalar/period/test_period.py]: can you parametrize next pass

File: pandas/tests/scalar/timedelta/test_arithmetic.py [hunk 4]
```diff
@@ -152,28 +163,17 @@ def test_td_sub_offset(self):
...
-        assert isinstance(result, Timedelta)
-        assert result == expected
-
-    def test_td_rsub_timedelta64(self):
-        td = Timedelta(10, unit='d')
-        expected = Timedelta(0, unit='ns')
-
-        result = td.to_timedelta64() - td
-        assert isinstance(result, Timedelta)
-        assert result == expected
```
Comment (jreback) [inline pandas/tests/scalar/timedelta/test_arithmetic.py]: parametrize next pass

File: pandas/tests/scalar/timestamp/test_comparisons.py [hunk 5]
```diff
@@ -12,6 +12,12 @@
 from pandas import Timestamp
 
 
+utc_objs = ['utc', utc, tzutc()]
```
Comment (jreback) [inline pandas/tests/scalar/timestamp/test_comparisons.py]: can you make this a fixture and put it in pandas/conftest.py. thought we had this already.

**Review by jreback — CHANGES_REQUESTED** [review]
just a couple of nitpicks, most can be done later.

File: pandas/tests/scalar/period/test_period.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/scalar/period/test_period.py]: Yah, I've got another branch focusing more closely on the scalar tests, will work this in there.

File: pandas/tests/scalar/timestamp/test_comparisons.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/tests/scalar/timestamp/test_comparisons.py]: sure.

You might be thinking of the datetime_tz_utc fixture which just provides the datetime.timezone.utc object (which seems like a really unnecessary fixture to me, especially seeing as how it is only used once)

File: pandas/core/arrays/datetimes.py [hunk 6]
```diff
@@ -222,6 +222,10 @@ def __new__(cls, values, freq=None, tz=None, dtype=None):
         # if dtype has an embedded tz, capture it
         tz = dtl.validate_tz_from_dtype(dtype, tz)
 
+        if isinstance(values, ABCSeries):
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimes.py]: Should this be done immediately? A `Series[datetime64[ns, tz]]` will not have a `.tz` attribute.

On this branch:

```python
In [16]: ser = pd.Series(pd.date_range('2017', periods=4, tz='UTC'))

In [17]: pd.core.arrays.DatetimeArrayMixin(ser).tz

In [18]: pd.core.arrays.DatetimeArrayMixin(ser._values).tz
Out[18]: <UTC>
```

I suspect we're never hitting this case (passing a series but not a dtype or `tz`), but fixing it now may save some future debugging.

Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: Good catch

**Comment by jbrockmendel** [comment]
Possible segfault in test_rank? I think this was changed s few days ago

**Comment by jbrockmendel** [comment]
gentle ping

**Review by TomAugspurger — APPROVED** [review]
+1. Will merge later today to give @jreback a chance to look if he wants.

**Review by jreback — APPROVED** [review]

**Pull request merged**